Adds Byond medals and leaderboards for megafauna kills.

Shadowlight213 opened the pull request:
Closes #19360
Same system, but it uses the byond hub medals as well as the scoreboard.

:cl:
add: Players will now receive an achievement for killing their first megafauna after this point.
add: Additionally, megafauna kills are now tracked on a hub page. Rack up those kills for first place!
/:cl:

Shadowlight213 committed:
Adds Byond medal and scoreboard support and tracking for megafauna kills.

RemieRichards commented on code/modules/admin/verbs/debug.dm:
```diff
@@ -777,3 +777,17 @@ var/global/list/g_fancy_list_of_types = null
+
+	message_admins("<span class='adminnotice'>[key_name_admin(src)] [global.medals_enabled ? "enabled" : "disabled"] the medal hub lockout.</span>")
+	feedback_add_details("admin_verb","TMH") // If...
+	log_admin("[key_name(src)] [global.medals_enabled ? "enabled" : "disabled"] the medal hub lockout.")
```
isn't this(and the above) backwards?

RemieRichards commented on code/modules/mob/living/simple_animal/hostile/megafauna/bubblegum.dm:
```diff
@@ -50,6 +50,8 @@ Difficulty: Hard
 	idle_vision_range = 5
 	loot = list(/obj/structure/closet/crate/necropolis/bubblegum)
 	var/charging = 0
+	medal_type = "Bubblegum"
```
no magic strings please.

Shadowlight213 replied:
Should I make a define for every one? I have to use strings with exact wording because of how medals work.

RemieRichards replied:
Of course.

RemieRichards commented on code/modules/mob/living/simple_animal/hostile/mining_mobs.dm:
```diff
@@ -1011,6 +1011,23 @@
 	qdel(gps)
 	. = ..()
 
+/mob/living/simple_animal/hostile/spawner/lavaland/death()
```
pretty sure you need to call ```...()``` for ```death()```

Jordie0608 commented on code/modules/mob/living/simple_animal/hostile/megafauna/megafauna.dm:
```diff
@@ -62,4 +67,116 @@
+			if(isnull(result))
+				global.medals_enabled = FALSE
+				log_game("MEDAL ERROR: Could not contact hub to award medal:[medal] player:[player.ckey]")
+				message_admins("Error! Failed to contact hub to award [medal] medal to [player.ckey]!")
```
I'm not sure this needs to announce to admins a failure in each of these procs.

Shadowlight213 replied:
this should only end up announcing once, as it will auto disable once it fails.

Jordie0608 replied:
I meant why announce it at all.

optimumtact replied:
Because a loud bug is a bug that gets fixed baka

Shadowlight213 replied:
Yup. Exactly.

MrStonedOne replied:
Make this a proc, database has a proc around connecting to the database, so this should be one too.

Jordie0608 commented on code/modules/mob/living/simple_animal/hostile/megafauna/megafauna.dm:
```diff
@@ -62,4 +67,116 @@
+				log_game("MEDAL ERROR: Could not contact hub to award medal:[medal] player:[player.ckey]")
+				message_admins("Error! Failed to contact hub to award [medal] medal to [player.ckey]!")
+			else if (result)
+				player << "<font color='green'><B>Achievement unlocked: [medal]!</B></font>"
```
Should use .greenannounce span class

optimumtact commented:
Does the ss13 hub support the medals?

Shadowlight213 committed:
defines and fixes.

Shadowlight213 commented:
@optimumtact No. I have already created a separate hub specifically for these.

Shadowlight213 committed:
span class

optimumtact commented:
Cool beans :bee:

Shadowlight213 committed:
adds ClearScore proc to remove a player from the leaderboard.

Shadowlight213 commented:
@optimumtact what kind of picture do you need?

optimumtact commented:
Just a picture of what a medal looks like, or where it's displayed, I've never seen one before.

KingDairy commented:
i still want my gloom medals to count for something though

Shadowlight213 commented:
Oh. well the medals are at http://www.byond.com/games/TgstationHub/TgstationStatsHub#tab=scores

The actual medals themselves can have custom icons. Kor was talking about making some.

optimumtact commented:
Looks blander than I was expecting

KorPhaeron commented:
Didn't even last one evening without an admin ignoring the GIANT RED TEXT not to spawn megafauna and messing up the stat tracking, I'm not sure achievements are viable long term.

optimumtact commented:
Wow relax son, they need to remake the hub because they committed the sin of calling it Tgstation

KorPhaeron commented:
Resetting the hub on a regular basis isn't gonna be viable though. 

And yeah I'm salty that the rule enforcers can't handle following a simple "don't do X" instruction.

optimumtact commented:
Just accept that you can't get perfect statistics on megafauna kills

KorPhaeron commented:
It's more the principle of following simple rules at this point.

Shadowlight213 commented:
It wasn't as bad as it seemed.
I had told them earlier that there was a debug verb to disable the medals updating, except it seems the verb didn't work.

Shadowlight213 committed:
Hopefully fixes the lockout failing.

jughu commented:
Shoulda made a exclusive version admins kor

MrStonedOne commented on config/config.txt:
```diff
@@ -187,6 +187,12 @@ TICKLAG 0.5
+
+## Password for the hub page
+MEDAL_HUB_PASSWORD
+
```
Needs to be more detailed.

Maybe an example?

Shadowlight213 replied:
I actually thought the same thing when I first was setting up the gloom hub page, but when you actually make it and look at the options, the hub address and password fields are clearly marked.
I can try adding an example for the address I guess. It would be ckey.hubname

Iamgoofball commented:
okay so:

admins who cant follow the big red text not to spawn megafauna need to be fired

stop being pussies about it and drop the hammer, if they cant pull their heads out of their asses long enough for that they can come back when they have some common sense

step 2:

have a var on megafauna that defaults to off and is set to on when the mob is spawned via lavaland's spawner code. As soon as an admin VV's it, the var is set to 0. The var is also VV locked.

If the var is enabled, it grants the achieves and counts to the scoreboard

if not, no go

GunHog commented:
I prefer Step 2. It is the best of both worlds. Boss monsters make for great events, it would be a shame to lose that because of this feature.

KorPhaeron commented:
They are not meant to be """events""" though they're supposed to be longer term goals for rd to work towards killing, though I realize they have failed at that purpose on sybil at least

GunHog commented:
How have they failed?

KorPhaeron commented:
People either solo them or ignore them for the most part. I've seen basil organize large raids though.

kilkun commented:
how does spawning them break tracking?

Jalleo commented:
Because Kor wants to know how many miners are killing on a per round basis.

Shadowlight213 commented:
Yeah. While unintended on my part initially, the scoreboard system works pretty well for tracking.

@jalleo the system would be more per player than per round I think.

optimumtact commented:
Ready?

MrStonedOne commented:
It is trivial to make it so only ones spawned organically track stats.

I'm not limiting how our admins do shit because somebody is too lazy to add a few checks.

Shadowlight213 commented:
Sure. I can add the checks easily. Eventually I need to create a more in depth debug verb for managing the various functions, but that's going to require a lot of work and I'm not sure if making the stats easy to directly modify would be a good thing.

Robustin commented:
More excuses to demote me eh?

But seriously, on a practical note, megafauna are currently something that like 2% of the station population engages with (as in fight, not just run as soon as you see them on the edge of your screen) and 1% actually know how to beat. 

It feels really absurd that despite all the work that went into making these cool megafauna, Kor and Shadow decided that these are basically going to be playthings for the 5-6 people who tryhard mining enough to take them on. The vast majority of the crew has never, and will never, experience megafauna because now we feel compelled to make an e-peen chart that those 5-6 people will care about. Everyone else either doesn't mine or just says "FUCK IT, RUN!" whenever they see one, they occasionally get smushed when one camps outside mining, and since only dozens of hours of trial and error on mining are the only way to learn how to fight these things - there's a huge barrier to anyone casually getting involved in this (ppsssst, most of our players are casual). 

All in all I applaud the effort here to dump a disproportionate amount of time and energy into an activity that becomes more and more niche with every PR involved (even I gave up trying to keep up with megafauna tactics after 2 more were added). I can't tell you how to spend your time, but this feels exceptionally pointless, even by SS13 Pull Request standards.

PKPenguin321 commented:
we should have a button next to the wiki/forums/changelog buttons to take you to the achievements hug

optimumtact commented:
The PR isn't pointless because it sets up the infrastructure and configuration for future medals. Just because the currently created set of medals only applies to a few things does not prevent someone for adding new medals for other more mundane achievements.

Durkel commented:
As literally the only guy who kills  bubblegum and colossus regularly I think this is stupid.

Shadowlight213 committed:
Adds admin_spawned atom var that is set when something is spawned by admin means or is var edited.
This can be bypassed by using debug-game if needed.
Stats will not be tracked for admin spawned atoms

Robustin commented:
It sounds like great infrastructure then, but its also the reason why admins can't incorporate any megafauna into events.

Like I said above, it is a goddamn tragedy you guys have put together the best hostile NPC's I've ever seen in SS13 and only a tiny fraction of the playerbase will ever see or meaningfully interact with them.

optimumtact commented:
![wherein robustin fails to perceive reality](https://file.house/VSY-.png)

KorPhaeron commented:
@Shadowlight213 

Is this done? It seems functional

Shadowlight213 commented:
it is. Just gotta do merge conflicts it seems

Shadowlight213 committed:
Fixes merge conflicts

Shadowlight213 commented:
@KorPhaeron yup.
I can definitely confirm functionality. every time I've seen a miner kill a megafauna the board updated accordingly.
